ENH: Add Wigner-Ville distribution

lndr opened the pull request:
#### Reference issue
This was mentioned in #3533, which was closed due to inaction.

#### What does this implement/fix?
Add Wigner-Ville distribution method for time-frequency analysis applications in signal processing.

#### Additional information
The method is implemented in a new module called `wigner`. I think about adding further methods related to the Wigner-Ville distribution, e.g. the Smoothed pseudo-Wigner–Ville distribution and the Gabor–Wigner transform. If desired, I can change the PR and add it to the `spectral` module.

rgommers commented on scipy/signal/tests/test_wigner.py:
```diff
@@ -0,0 +1,24 @@
+        t, f, wv = wigner.wvd(x_re + 1j * x_im, fs)
+        max_idx = np.unravel_index(np.argmax(wv), wv.shape)
+        assert np.allclose(fc, f[max_idx[0]], atol=f[1] - f[0])
+        assert np.allclose(tc, t[max_idx[1]], atol=t[1] - t[0])
```
Using `assert_allclose` from `numpy.testing` would be better.

lndr replied:
Done.

rgommers commented on scipy/signal/wigner.py:
```diff
@@ -0,0 +1,95 @@
+    Examples
+    --------
+    >>> import matplotlib.pyplot as plt
+    >>> import numpy as np
```
this import isn't needed, it's assumed in all our examples

lndr replied:
Done.

rgommers commented on scipy/signal/wigner.py:
```diff
@@ -0,0 +1,95 @@
+__all__ = ['wvd']
+
+
+def wvd(x, fs=1.0, resolution=1, win_size=None):
```
Could you think of a more readable name? This is too Matlab-y. Maybe just `wigner_ville`?

lndr replied:
The less Matlab-y the better, especially because thier `wvd` is a mix of different methods related to the Wigner-Ville distribution. `wigner_ville` is fine with me.

rgommers commented on scipy/signal/wigner.py:
```diff
@@ -0,0 +1,95 @@
+    resolution : int, optional
+        Resolution, number of samples between windows. Defaults to 1.
+    win_size : int, optional
+        Window size, length of frequency axis. Defaults to len(x) // 2.
```
surround the code expression by double backticks

lndr replied:
Done.

rgommers commented on scipy/signal/wigner.py:
```diff
@@ -0,0 +1,95 @@
+def wvd(x, fs=1.0, resolution=1, win_size=None):
+    """
+    Compute the Wigner-Ville distribution.
+
```
Could you add one paragraph here about what the Wigner-Ville distribution is/does?

lndr replied:
Done.

rgommers commented on scipy/signal/__init__.py:
```diff
@@ -307,6 +307,7 @@
 from ._savitzky_golay import savgol_coeffs, savgol_filter
 from .spectral import *
 from .wavelets import *
+from .wigner import *
```
Just importing the single function by name would be preferred for new code.

lndr replied:
Done.

rgommers commented on scipy/signal/wigner.py:
```diff
@@ -0,0 +1,95 @@
+        Sampling frequency of the `x` time series. Defaults to 1.0.
+    resolution : int, optional
+        Resolution, number of samples between windows. Defaults to 1.
+    win_size : int, optional
```
This parameter is called `window_size` in other functions.

lndr replied:
Done.

rgommers reviewed (changes requested):
Thanks @lndr!

I did an initial cursory review. Getting the API right is important, because we cannot change it later. I didn't look at the algorithm in detail.

Could you send an email to the scipy-dev mailing list proposing to add this function? That's what we do for all new functionality, and hopefully that'll also attract a reviewer who has a need for this function.

lndr commented:
One general remark: Also it is called a distribution, the WVD is a tool mostly used for time-frequency signal analysis and should thus be labeled `scipy.signal`.

BvB93 commented on scipy/signal/wigner.py:
```diff
@@ -0,0 +1,99 @@
+    >>> plt.ylim([0, 6])
+    >>> plt.show()
+    """
+    if not window_size:
+        window_size = len(x) // 2
```
```suggestion
    x = np.asarray(x)
    if not window_size:
        window_size = len(x) // 2
```
With the large variety of different array-likes that's out there, there is no guarantee that `x` is a `Sized` object, 
even if it does have the correct dimensions. Explicitly calling `np.asarray` beforehand should resolve these 
(somewhat niche) cases.

BvB93 replied:
``` diff
- if not window_size:
+ if window_size is not None:
```
Unless someone is willing to argue that `None` and `0` should be treated as equivalent, 
I'd also suggest to perform an explicit `is None` check rather than truth-testing.

lndr replied:
> With the large variety of different array-likes that's out there, there is no guarantee that `x` is a `Sized` object,
> even if it does have the correct dimensions. Explicitly calling `np.asarray` beforehand should resolve these
> (somewhat niche) cases.

You never know what people come up with. Done.


> ```diff
> - if not window_size:
> + if window_size is not None:
> ```
No need to argue, but I guess it's supposed to be `if window_size is None`.

BvB93 commented on scipy/signal/wigner.py:
```diff
@@ -0,0 +1,99 @@
+__all__ = ['wvd']
+
+
+def wigner_ville(x, fs=1.0, resolution=1, window_size=None):
```
Would you be willing to add annotations for the new function?
It's signature seems to be rather straightforward.
```suggestion
def wigner_ville(
    x: npt.ArrayLike, 
    fs: float = 1.0, 
    resolution: int = 1, 
    window_size: None | int = None,
) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
```
Note that this will also require the following imports:
``` python
from __future__ import annotations

from typing import Tuple, TYPE_CHECKING

if TYPE_CHECKING:
    import numpy.typing as npt
```

lndr replied:
I like that. I just had not seen type hints in any of the other scipy methods I used as a reference.

czgdp1807 commented:
This is an interesting work. I am not expert in quantum mechanics (just started out with some introductory lectures and have some intuitive knowledge of this field). I was reading [Wigner quasiprobability distribution](https://en.wikipedia.org/wiki/Wigner_quasiprobability_distribution) which defines the distribution in terms of position and momentum of a particle having wave function `f`. May be we can have this definition in `scipy.stats`. In the Mathematical properties the wikipedia article also defines probability distributions for `x`, `p` by marginalising `W(x, p)`. Please let me know if you are also doing the same thing.

lndr commented:
> This is an interesting work. I am not expert in quantum mechanics (just started out with some introductory lectures and have some intuitive knowledge of this field). I was reading [Wigner quasiprobability distribution](https://en.wikipedia.org/wiki/Wigner_quasiprobability_distribution) which defines the distribution in terms of position and momentum of a particle having wave function `f`. May be we can have this definition in `scipy.stats`. In the Mathematical properties the wikipedia article also defines probability distributions for `x`, `p` by marginalising `W(x, p)`. Please let me know if you are also doing the same thing.

I am also not into quantum mechanics, the proposed implementation is more aimed at usage in signal analysis (time-frequency-analysis to be specific), hence the placement in scipy.signal. The appropriate wiki article would be https://en.wikipedia.org/wiki/Wigner_distribution_function. Unfortunately, naming is all over the place, not only on wikipedia. So, while the basic idea is similar, the applications are so different that I wouldn't say that they are the same thing.

scimax commented:
Hello everyone, I'd like to see this feature in scipy! Thanks for the hard work @lndr ! Is there anything I could do to support this PR?

miguelfs commented:
Shouldn't the Gabor transform be a distinct function by itself, just like [STFT](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.stft.html) and [CWT](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.cwt.html)?

lndr commented:
> Shouldn't the [Gabor-Wigner transform](https://en.wikipedia.org/wiki/Gabor%E2%80%93Wigner_transform) be a distinct function by itself, just like [STFT](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.stft.html) and [CWT](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.cwt.html)?

It should, but it may be possible to reuse some of the code written for the Wiegner-Ville distribution. It could thus be placed in the same module.

tupui reviewed (commented):
Some fly-by comments to help move this forward.

I also see this was sent to the mailing list: https://mail.python.org/archives/list/scipy-dev@python.org/message/ZOZHIMZ6CYK3Q4K4HT6SK7WZHIDWKINY/
No reply means no complains. Also seems like some people expressed interest here.

lndr commented on scipy/signal/__init__.py:
```diff
@@ -326,6 +327,7 @@
 from ._signaltools import *
 from ._savitzky_golay import savgol_coeffs, savgol_filter
+from .wigner import wigner_ville
 from ._spectral_py import *
```
Fixed this in 14bdfbb.

lndr replied:
Thank you @tupui for your remarks. I almost considered this PR to be dead. 
I hope I have resolved all your remarks accordingly. There seem to have been quite some changed to the general conventions since I opened the PR.

tupui commented on scipy/signal/tests/test_wigner.py:
```diff
@@ -0,0 +1,25 @@
+from numpy.testing import assert_allclose
+from scipy.signal import gausspulse
+
+from scipy.signal import wigner
```
```suggestion
from scipy.signal import wigner_ville
```

tupui commented on scipy/signal/tests/test_wigner.py:
```diff
@@ -0,0 +1,25 @@
+
+class TestWigner:
+    def test_wigner_ville_shapes(self):
+        t, f, wv = wigner.wigner_ville(np.sin(np.linspace(0, 100, 1024)),
```
```suggestion
        t, f, wv = wigner_ville(np.sin(np.linspace(0, 100, 1024)),
```

tupui commented on scipy/signal/tests/test_wigner.py:
```diff
@@ -0,0 +1,25 @@
+        tc = 512 / fs
+        t = np.arange(0, 1024) / fs
+        x_re, x_im = gausspulse(t - tc, fc, bw=0.2, retquad=True)
+        t, f, wv = wigner.wigner_ville(x_re + 1j * x_im, fs)
```
```suggestion
        t, f, wv = wigner_ville(x_re + 1j * x_im, fs)
```

rgommers commented:
@larsoner we discussed this in the community meeting today, and figured it will make sense to move it to `_spectral_py.py`. Does that sound good? Also, maybe you'd be interested to have a look at this?

tylerjereddy commented:
@larsoner @lndr I'm hoping to branch for SciPy `1.12.0` on December 6th, in case this PR has a chance of getting revived before then.

lndr commented:
I would hope so. When we talked about in in the community meeting a year ago, the only thing missing, aside from maybe moving it to the spectral module, was for someone to review the math.

DietBru commented:
I found some time to investigate this WVD implementation a bit: The following statements should be read in the light of @rgommer's [comment](https://github.com/scipy/scipy/pull/14141#pullrequestreview-682353436) as being thought food for reasoning about the api a little more:

1. The WVD being a pseudo power density is its distinguishing feature among time-frequency representations. This means its marginals, which can be determined by (numerically) integrating over time or frequency, are the power spectral density (PSD) and the instantaneous power (i.e., `|x(t)|²`). It would be nice if there would be unit tests as well an example in the docstring verifying that. (For background information on the PSD of a sampled signal, you are more than welcome to review PR #19488 😁).
2. It seems that a scaling factor is missing. It can be verified by investigating the marginals. 
3. The provided function only determines the WVD for non-negative frequencies. The negative part is needed for complex-valued signals. The [fft_mode](https://scipy.github.io/devdocs/reference/generated/scipy.signal.ShortTimeFFT.fft_mode.html#scipy.signal.ShortTimeFFT.fft_mode) property of the [ShortTimeFFT](https://scipy.github.io/devdocs/reference/generated/scipy.signal.ShortTimeFFT.html) shows an example on how to parametrize the FFT for different use cases.
4. IMHO, it would be straightforward to extend this function to be able to calculate a cross-WVD.
5. One of the lessons of new [ShortTimeFFT](https://scipy.github.io/devdocs/reference/generated/scipy.signal.ShortTimeFFT.html) implementation was that the signal's borders need to be carefully considered (the legacy [stft](https://scipy.github.io/devdocs/reference/generated/scipy.signal.stft.html#scipy.signal.stft)/[istft](https://scipy.github.io/devdocs/reference/generated/scipy.signal.istft.html#scipy.signal.istft) were a bit intransparent in that regard):
   * My guess is the time domain of the WVD is larger than that of the input signal—at least when the inverse WVD should be able to reconstruct the original signal.   
   * It might be nice to be able to assemble WVDs of sequential signal chunks into a large WVD (the cross-WVD could be helpful in that regard).
6. It would be nice to have the inverse WVD (with the same api of course).
7. The WVD is not guaranteed to be non-negative (the reason for naming it "pseudo-density"). Smoothing the WVD with an appropriate Gaussian kernel removes the negative parts at the expense of resolution (e.g., consult Chap. 4.4 of [Gröechnig's book](https://doi.org/10.1007/978-1-4612-0003-1)). Hence, having a smoothed WVD would enhance the usability for real-world problems a lot.

lndr commented:
Thank you for the review @DietBru! Nice to see some activity in the PR.

1: Such a test case would indeed be quite sensible. I'll have a look into it, but wouldn't you say that having it as an example would be quite a lot of code?
2: I'll look that up.
3: Seems sensible to include a feature like fftmode. I'll have a look at that to.
5.1: It should be len(x) * 2 - 1 because of the autocorrelation, right? Might be a good idea to keep it that way. 
5.2: That seems like quite an advanced feature, maybe a bit out of the scope of this PR?
4,6,7: I've considered the additional functionality you proposed (IWVD, cross-WVD PWVD, etc.) and concluded to first implement only a basic function to judge if there is interest in such features. 

Thanks again for the feedback. Unfortunately, I'm quite busy at the moment and it was quite some time since it wrote that code. I'll try to work on it, but any support is welcome.

tylerjereddy commented:
It looks like there is some good feedback here now. Perhaps if the discussion remains active for a little bit longer folks could agree on keeping the scope to some reasonable target. The requests for adding more tests might be considered important enough to delay a merge though.

I'll probably bump the milestone to `1.13.0` here, but note that `1.13.0` is likely going to be less than six months away, because it will aim to support NumPy `2.0.0` when it comes out, so there may be another medium-term opportunity to get this in if things align.

lndr commented:
I finally found some time to work on the PR again. I added an additional test following the suggestion by @DietBru, which tests for the marginals. While implementing that, I noticed some problems with the algorithm and looked up the implementation by Bradford that inspired it. I added an additional reference to the documentation for that, which is freely available at [Caltech](https://thesis.library.caltech.edu/4689/). @DietBru already suspected that there is something wrong with the scaling, but I found the problem lying somewhat deeper. The implementation at hand, while quite efficient, achieves this by requiring the input signal to be an analytic signal. It does unfortunately not represent the general case of the Wigner-Ville distribution.

I'm not quite sure one would continue from here. There are suggestions for [efficient, general implementations](https://www.sciencedirect.com/science/article/pii/S1063520313000055) and other python implementation in smaller packages, for example in [pytftb](https://github.com/jaidevd/pytftb). If we are to continue, I would suggest starting with the cross Wigner-Ville distribution, then implementing the ordinary WVD as a special cases of that, and then implement pseudo-WVDs etc. I guess we would do this in a different PR then, because not much of the current code would survive.
However, interest in making the WVD and its derivatives a part of scipy has not been on the high side in the two and a half years this PR has been oben. So before we/I start pouring more work in this project, I'd like to ask everyone how contributed thus far: Do we want the Wigner-Ville distribution in scipy or not?

lucascolley commented:
> I'm not quite sure one would continue from here. There are suggestions for [efficient, general implementations](https://www.sciencedirect.com/science/article/pii/S1063520313000055) and other python implementation in smaller packages, for example in [pytftb](https://github.com/jaidevd/pytftb). If we are to continue, I would suggest starting with the cross Wigner-Ville distribution, then implementing the ordinary WVD as a special cases of that, and then implement pseudo-WVDs etc. I guess we would do this in a different PR then, because not much of the current code would survive.
However, interest in making the WVD and its derivatives a part of scipy has not been on the high side in the two and a half years this PR has been oben. So before we/I start pouring more work in this project, I'd like to ask everyone how contributed thus far: Do we want the Wigner-Ville distribution in scipy or not?

@larsoner any opinion in light of this?

ljbkusters commented:
> I finally found some time to work on the PR again. I added an additional test following the suggestion by @DietBru, which tests for the marginals. While implementing that, I noticed some problems with the algorithm and looked up the implementation by Bradford that inspired it. I added an additional reference to the documentation for that, which is freely available at [Caltech](https://thesis.library.caltech.edu/4689/). @DietBru already suspected that there is something wrong with the scaling, but I found the problem lying somewhat deeper. The implementation at hand, while quite efficient, achieves this by requiring the input signal to be an analytic signal. It does unfortunately not represent the general case of the Wigner-Ville distribution.
> 
> I'm not quite sure one would continue from here. There are suggestions for [efficient, general implementations](https://www.sciencedirect.com/science/article/pii/S1063520313000055) and other python implementation in smaller packages, for example in [pytftb](https://github.com/jaidevd/pytftb). If we are to continue, I would suggest starting with the cross Wigner-Ville distribution, then implementing the ordinary WVD as a special cases of that, and then implement pseudo-WVDs etc. I guess we would do this in a different PR then, because not much of the current code would survive. However, interest in making the WVD and its derivatives a part of scipy has not been on the high side in the two and a half years this PR has been oben. So before we/I start pouring more work in this project, I'd like to ask everyone how contributed thus far: Do we want the Wigner-Ville distribution in scipy or not?

I think the Wigner-Ville Distribution is one of those things nobody knows about untill they need it. I found myself in this situation a couple of years ago and turned to scipy to find an implementation of the WVD, and then being stumped as to why there was no implementation of this. In the end I just created my own version of a pseudo-wigner ville distribution. Not to toot my own horn or anything but I have a [repo](https://github.com/ljbkusters/python-wigner-distribution) implementing the Wigner distribution without a for loop using only `numpy` and `scipy` methods, which I assume would be a free speedup over a Python loop, although I have no metrics on how they compare both time-wise and in qualitatively (`pytftb` also uses python for loops from what I could gather). At the time I created it with the goal of sometime opening a PR to scipy and merging it but I think approaching it from the general implementation and implementing other parts as special cases makes sense. Then again, if it takes too much work to solve the general problem while we only need a pseudo wigner ville distribution most of the time, that might be overkill.

lndr commented:
> > I finally found some time to work on the PR again. I added an additional test following the suggestion by @DietBru, which tests for the marginals. While implementing that, I noticed some problems with the algorithm and looked up the implementation by Bradford that inspired it. I added an additional reference to the documentation for that, which is freely available at [Caltech](https://thesis.library.caltech.edu/4689/). @DietBru already suspected that there is something wrong with the scaling, but I found the problem lying somewhat deeper. The implementation at hand, while quite efficient, achieves this by requiring the input signal to be an analytic signal. It does unfortunately not represent the general case of the Wigner-Ville distribution.
> > I'm not quite sure one would continue from here. There are suggestions for [efficient, general implementations](https://www.sciencedirect.com/science/article/pii/S1063520313000055) and other python implementation in smaller packages, for example in [pytftb](https://github.com/jaidevd/pytftb). If we are to continue, I would suggest starting with the cross Wigner-Ville distribution, then implementing the ordinary WVD as a special cases of that, and then implement pseudo-WVDs etc. I guess we would do this in a different PR then, because not much of the current code would survive. However, interest in making the WVD and its derivatives a part of scipy has not been on the high side in the two and a half years this PR has been oben. So before we/I start pouring more work in this project, I'd like to ask everyone how contributed thus far: Do we want the Wigner-Ville distribution in scipy or not?
> 
> I think the Wigner-Ville Distribution is one of those things nobody knows about untill they need it. I found myself in this situation a couple of years ago and turned to scipy to find an implementation of the WVD, and then being stumped as to why there was no implementation of this. In the end I just created my own version of a pseudo-wigner ville distribution. Not to toot my own horn or anything but I have a [repo](https://github.com/ljbkusters/python-wigner-distribution) implementing the Wigner distribution without a for loop using only `numpy` and `scipy` methods, which I assume would be a free speedup over a Python loop, although I have no metrics on how they compare both time-wise and in qualitatively (`pytftb` also uses python for loops from what I could gather). At the time I created it with the goal of sometime opening a PR to scipy and merging it but I think approaching it from the general implementation and implementing other parts as special cases makes sense. Then again, if it takes too much work to solve the general problem while we only need a pseudo wigner ville distribution most of the time, that might be overkill.

Thanks for your input. Judging from the responses to this PR, interest in having the WVD in SciPy does not seem to be high, unfortunately.
I had a look at your implementation. Did you develop the algorithm yourself? The problem with there implementation I used is that we can’t really track it to its origin anymore. Maybe we can use your approach but keep the interface already in the PR which we already had some discussion about. 
While I concur that you’d use a smoothed pseudo WVD most of the time, the plain WVD has its applications as well, so I’d suggest at least we would need these two, don’t you think?

ljbkusters commented:
Yes, I developed it myself, based on the [paper](https://www.researchgate.net/publication/257291758_The_Wigner_distribution-A_tool_for_time-frequency_signal_analysis-Part_II_Discrete_time_signals) I link in the docstring. Specifically equations 2.10 and 2.11. The implementation is just some manipulation of indices to get the desired terms `(n+k)` and `(n-k)` per row of data in the `wigner_integrand` matrix and then let the FFT do the heavy lifting. I think my implementation easily extends to the cross-wigner distribution (defined over `f` and `g`) by the way. On the topic of the smoothed wigner distribution, although I agree that it is useful I'm not sure if it should be a part of SciPy directly.

lndr commented:
> Yes, I developed it myself, based on the [paper](https://www.researchgate.net/publication/257291758_The_Wigner_distribution-A_tool_for_time-frequency_signal_analysis-Part_II_Discrete_time_signals) I link in the docstring. Specifically equations 2.10 and 2.11. The implementation is just some manipulation of indices to get the desired terms `(n+k)` and `(n-k)` per row of data in the `wigner_integrand` matrix and then let the FFT do the heavy lifting. I think my implementation easily extends to the cross-wigner distribution (defined over `f` and `g`) by the way. On the topic of the smoothed wigner distribution, although I agree that it is useful I'm not sure if it should be a part of SciPy directly.

That sounds promising. If it is okay for you I'll try to modify the code in the PR with your algorithm, keeping the interface, when I find some time to spare. Afterwards, you can review it.

lndr commented:
> > That sounds promising. If it is okay for you I'll try to modify the code in the PR with your algorithm, keeping the interface, when I find some time to spare. Afterwards, you can review it.
> 
> Sure.
> 
> I would like to add that I could also contribute in my own spare time. I'm not quite sure what the most appropriate way to do that is, should I fork your fork?

I just invited you as a collaborator on my fork of scipy. Since I find it hard to work on this one at the moment, you can push to the wigner branch. Some thoughts I had since it last worked on it:
* I think we should **not** compute the analytic signal of the input inside the wigner method, even if it is somewhat of a default case. People should know what they are doing when they use the method, and there are quite some use cases where we compute the WVD of a complex signal.
* If you like you could also implement the cross-wigner distribution, with the same interface but two input signals, and just make the WVD as a special case of that. 
* Feel free to implement further tests. I find it quite challenging to come up with good test cases.
* Just have a look a the discussion above. Maybe there are some points it missed ;)

Thanks again for digging this up. I was about to call it off and close the PR.
